public/app/plugins/panel/gauge/Thresholds.tsx

```diff
@@ -0,0 +1,306 @@
+import React, { PureComponent } from 'react';
+import classNames from 'classnames/bind';
+import { ColorPicker } from 'app/core/components/colorpicker/ColorPicker';
+import { OptionModuleProps } from './module';
+import { Threshold } from 'app/types';
+
+interface State {
+  thresholds: Threshold[];
+}
+
+enum BasicGaugeColor {
+  Green = 'rgba(50, 172, 45, 0.97)',
+  Orange = 'rgba(237, 129, 40, 0.89)',
+  Red = 'rgb(212, 74, 58)',
+}
+
+export default class Thresholds extends PureComponent<OptionModuleProps, State> {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      thresholds: this.props.options.thresholds || [
+        { index: 0, label: 'Min', value: 0, canRemove: false, color: BasicGaugeColor.Green },
+        { index: 1, label: 'Max', value: 100, canRemove: false },
+      ],
+    };
+  }
+
+  onAddThreshold = index => {
+    const { thresholds } = this.state;
+
+    const newThresholds = thresholds.map(threshold => {
+      if (threshold.index >= index) {
+        threshold = { ...threshold, index: threshold.index + 1 };
+      }
+
+      return threshold;
+    });
+
+    // Setting value to a value between the new threshold.
+    const value = newThresholds[index].value - (newThresholds[index].value - newThresholds[index - 1].value) / 2;
+
+    this.setState(
+      {
+        thresholds: this.sortThresholds([
+          ...newThresholds,
+          { index: index, label: '', value: value, canRemove: true, color: BasicGaugeColor.Orange },
+        ]),
+      },
+      () => this.updateGauge()
+    );
+  };
+
+  onRemoveThreshold = threshold => {
+    this.setState(
+      prevState => ({
+        thresholds: prevState.thresholds.filter(t => t !== threshold),
+      }),
+      () => this.updateGauge()
```

Comment:
Nice! I tend to forget about the setState callback and use await instead. This is much cleaner imho.